exec/java-exec/src/main/java/org/apache/drill/exec/rpc/user/UserClient.java

```diff
@@ -88,22 +129,183 @@ public void submitQuery(UserResultsListener resultsListener, RunQuery query) {
     send(queryResultHandler.getWrappedListener(resultsListener), RpcType.RUN_QUERY, query, QueryId.class);
   }
 
-  public void connect(RpcConnectionHandler<ServerConnection> handler, DrillbitEndpoint endpoint,
-                      UserProperties props, UserBitShared.UserCredentials credentials) {
+  public CheckedFuture<Void, RpcException> connect(DrillbitEndpoint endpoint, DrillProperties parameters,
+                                                   UserCredentials credentials) {
+    final FutureHandler handler = new FutureHandler();
     UserToBitHandshake.Builder hsBuilder = UserToBitHandshake.newBuilder()
         .setRpcVersion(UserRpcConfig.RPC_VERSION)
         .setSupportListening(true)
         .setSupportComplexTypes(supportComplexTypes)
         .setSupportTimeout(true)
         .setCredentials(credentials)
-        .setClientInfos(UserRpcUtils.getRpcEndpointInfos(clientName));
+        .setClientInfos(UserRpcUtils.getRpcEndpointInfos(clientName))
+        .setSaslSupport(SaslSupport.SASL_AUTH)
+        .setProperties(parameters.serializeForServer());
+    this.properties = parameters;
+
+    connectAsClient(queryResultHandler.getWrappedConnectionHandler(handler),
+        hsBuilder.build(), endpoint.getAddress(), endpoint.getUserPort());
+    return handler;
+  }
+
+  /**
+   * Check (after {@link #connect connecting}) if server requires authentication.
+   *
+   * @return true if server requires authentication
+   */
+  public boolean serverRequiresAuthentication() {
+    return supportedAuthMechs != null;
+  }
+
+  /**
+   * Returns a list of supported authentication mechanism. If called before {@link #connect connecting},
+   * returns null. If called after {@link #connect connecting}, returns a list of supported mechanisms
+   * iff authentication is required.
+   *
+   * @return list of supported authentication mechanisms
+   */
+  public List<String> getSupportedAuthenticationMechanisms() {
```

Comment:
Will defer to future. For now, I will remove this method, and make auth part of `connect`.